Remove side effects to improve Ivy tree shaking in Closure

dgp1130 opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
N/A - No easy way to verify that Closure tree shakes the relevant functions.
- [X] Docs have been added / updated (for bug fixes / features)
No docs to update.

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Bugfix

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Closure does not tree shake unused components built with Ivy.

Issue Number: http://b/149874196, http://b/149874638.

## What is the new behavior?

Closure will now tree shake unused components built with Ivy.

## Does this PR introduce a breaking change?

- [X] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Closure is failing to tree shake components because it is detecting side effects in the definition which it is unable to remove. Most of these are fixed by adding `noSideEffects()` to the relevant calls and generally tweaking things so Closure has a better understanding of what can be stripped if not used.

I made separate commits so that each solves its own side effect problem. Let me know if I should squash any/all of these commits together. Also the `NgOnChangesFeature` touches the compiler as well as Core, so I'm not sure what to put in the commit message. Let me know if I should change that somehow.

dgp1130 committed:
refactor(core): update `noSideEffects()` to return the result of the inner function

This is useful for propagating return values without them being converted to a string. It still provides the same guarantees to Closure, which will assume that the function invoked is pure and can be tree-shaken accordingly.

dgp1130 committed:
fix(core): add `noSideEffects()` to `make*Decorator()` functions

This causes all the `make*Decorator()` functions to be considered pure and to be eligible for associated tree shaking by Closure.

dgp1130 committed:
fix(core): add `noSideEffects()` to `ɵɵdefineComponent()`

This marks the function are "pure" and eligible to be tree shaken by Closure. Without this, initializing `ngDevMode` is considered a side effect which prevents this function from being tree shaken and also any component which calls it.

dgp1130 committed:
fix(core): remove side effects from `ɵɵNgOnChangesFeature()`

`ɵɵNgOnChangesFeature()` would set `ngInherit`, which is a side effect and also not necessary. This was pulled out to module scope so the function itself can be pure. Since it only curries another function, the call is entirely unnecessary. Updated the compiler to only generate a reference to this function, rather than a call to it, and removed the extra curry indirection.

dgp1130 committed:
fix(core): remove side effects from `ɵɵgetInheritedFactory()`

`ɵɵgetInheritedFactory()` is called from generated code for a component which extends another class. This function is detected by Closure to have a side effect and is not able to tree shake the component as a result. Marking it with `noSideEffects()` tells Closure it can remove this function under the relevant tree shaking conditions.

IgorMinar commented on packages/core/src/util/closure.ts:
```diff
@@ -15,6 +15,6 @@
-}
\ No newline at end of file
+export function noSideEffects<T>(fn: () => T): T {
+  return {toString: fn}.toString() as unknown as T;
```
thanks for using `unknown` instead of more commonly (but now incorrectly) used `any` 👍

IgorMinar reviewed (approved):
this looks great @dgp1130 !! it needs just a bit more polish but structurally all these changes look solid!

also thanks for documenting the reasoning behind each change and breaking them up into individual commits.

some small things to fix:

```
FAIL: Commit 11a6a588e6ceaed9eef24bd05b886de23ced47ba uncompressed bundle fell below expected size by 500 bytes or >1% (expected: 175498, actual: 170618).
If this is a desired change, please update the size limits in file '/home/circleci/ng/integration/../integration/_payload-limits.json'.
```
- [x] please update `integration/_payload-limits.json` with the new decreased value (170618) for the closure limit.
- [x] please run `yarn gulp format` to reformat the code changes
- [x] please run `yarn bazel run --config=ivy //packages/core/test/bundling/cyclic_import:symbol_test.accept` to update golden files
- [x] please run g3 presubmit once github CI is green

GLOBAL APPROVAL unless further significant modifications are made.

dgp1130 commented:
Addressed your comments Igor, thanks for the direction on updating goldens. Applied the golden updates to the latest commit. I had to force g3 to pass as the only failures were pre-existing and flaky.

Should be ready to merge.
